unit_tests/ut_h_upgrade_htest.amlg

```diff
@@ -0,0 +1,215 @@
+		method "train"
+		payload
+			(assoc
+				features (list "x" "y" "w")
+				cases
+					(list
+						(list 4 4 4)
+						(list 5 5 5)
+					)
+			)
+	))
+
+	(store_entity "./htest.amlg" "howso" "amlg" (false) {flatten (true)})
+
+	;reset howso entity to be blank trainee
+	(call (load "unit_test_howso.amlg")
+		(assoc name "ut_h_update_htest.amlg" skip_init (true) do_return_validation (false))
+	)
+
+	(call_entity "howso" "upgrade_trainee" (assoc
+		trainee "htest"
+		root_filepath "../"
+		trainee_amlg_filepath "./"
+	))
+
+	(print (call_entity "howso" "get_hierarchy"))
+
+	;Checking feature attributes
+	(assign (assoc
+		result
+			(call_entity "howso" "execute_on_subtrainee" (assoc
+				path ["A" "b" "c2"]
+				method "get_feature_attributes"
+				payload {}
+			))
+	))
+	(call keep_result_payload)
+	(print "Feature Attributes of c2 preserved: ")
+	(call assert_same (assoc
+		obs (zip (indices result))
+		exp (zip ["x" "y" "w"])
+	))
+
+	(assign (assoc
+		result
+			(call_entity "howso" "execute_on_subtrainee" (assoc
+				path ["A" "b"]
+				method "get_feature_attributes"
+				payload {}
+			))
+	))
+	(call keep_result_payload)
+	(print "Feature Attributes of b preserved: ")
+	(call assert_same (assoc
+		obs (zip (indices result))
+		exp (zip ["x" "y"])
+	))
+
+	(assign (assoc
+		result (call_entity "howso" "get_feature_attributes" {})
+	))
+	(call keep_result_payload)
+	(print "Feature Attributes of root preserved: ")
+	(call assert_same (assoc
+		obs (zip (indices result))
+		exp (zip ["a" "b" "c"])
+	))
+
+	(call exit_if_failures {msg "Feature Attributes not preserved correctly."})
+
+	;Checking Parent IDs
+	(assign (assoc
+		result
+			(call_entity "howso" "execute_on_subtrainee" (assoc
+				path ["A" "b"]
+				method "debug_label"
+				payload {label "!parentId"}
+			))
+	))
+	(print "Parent of b: ")
+	(call assert_same (assoc
+		obs result
+		exp "A"
+	))
+
+	(assign (assoc
+		result
+			(call_entity "howso" "execute_on_subtrainee" (assoc
+				path ["A" "b" "c1"]
+				method "debug_label"
+				payload {label "!parentId"}
+			))
+	))
+	(print "Parent of c1: ")
+	(call assert_same (assoc
+		obs result
+		exp "b"
+	))
+
+	(call exit_if_failures {msg "Parent Ids not preserved correctly."})
+
+	;Checking num cases
+	(assign (assoc
+		result
+			(call_entity "howso" "execute_on_subtrainee" (assoc
+				path ["A" "b"]
+				method "get_num_training_cases"
+				payload {}
+			))
+	))
+	(call keep_result_payload)
+	(print "Num training cases in b: ")
+	(call assert_same (assoc
+		obs (get result "count")
+		exp 3
+	))
+
+	(assign (assoc
+		result
+			(call_entity "howso" "execute_on_subtrainee" (assoc
+				path ["A" "b" "c2"]
+				method "get_num_training_cases"
+				payload {}
+			))
+	))
+	(call keep_result_payload)
+	(print "Num training cases in c2: ")
+	(call assert_same (assoc
+		obs (get result "count")
+		exp 2
+	))
+
+	(call exit_if_failures {msg "Cases not preserved correctly."})
+
+	(if (= (system "os") "Windows")
```

Comment:
outside the scope the scope of this task, but we should add an optional parameter to 'exit_if_failures' in unit_test to call a cleanup method on exit when failed, then slap a cleanup label on this `if` statement, that way all the exit_if_failures in this test will still delete the htest.amlg file even if the test fails.

Reply:
sure, done